### PR TITLE
Cherrypick commits related to `placeHolderVar`

### PR DESCRIPTION
**Pull request by dhanashreek89**
This is a cherry-pick of upstream commits that introduced `PlaceHolder` variable mechanism.

e6ae3b5  Introduces PlaceHolderVars
31468d0, c59d8dd, 2bdd765, dc9cc88 refine the commit above.

### CODE REVIEW

File: src/test/regress/expected/subselect_gp.out [hunk 1]
```diff
@@ -1625,7 +1623,7 @@ EXPLAIN select count(*) from
  Aggregate  (cost=752.98..752.99 rows=1 width=8)
    ->  Gather Motion 3:1  (slice2; segments: 3)  (cost=752.91..752.96 rows=1 width=8)
          ->  Aggregate  (cost=752.91..752.92 rows=1 width=8)
-               ->  Hash Join  (cost=414.79..728.02 rows=3319 width=4)
```
Comment (hlinnaka) [inline src/test/regress/expected/subselect_gp.out]: Something happened to the width estimates here. I haven't looked at the query, but 0 doesn't seem right.

**Comment by dhanashreek89** [comment]
@hlinnaka Right, it was a straight forward back-port. No GPDB specific changes except that these walkers have different skeleton in GPDB than Upstream:
`pull_varnos_of_level`, `contain_var_reference`,  `find_minimum_var_level`

Hence changes in these are a little bit different than upstream.

File: src/test/regress/expected/subselect_gp.out (hunk 1 above)
Comment (bhuvnesh2703) [inline src/test/regress/expected/subselect_gp.out]: We took a look at the upstream results, there as well we see width as 0.
If the operator on top of HASHJOIN is not using any projection from HASHJOIN, in that case we are setting the width to 0. Plan from postgres 8.4
```
postgres=# EXPLAIN select count(*) from                                                 (select 1 from tenk1 a                                                                   where unique1 IN (select distinct hundred from tenk1 b)) ss;
                                    QUERY PLAN
----------------------------------------------------------------------------------
 Aggregate  (cost=37.38..37.39 rows=1 width=0)
   ->  Hash Join  (cost=20.12..37.01 rows=145 width=0)
         Hash Cond: (a.unique1 = b.hundred)
         ->  Seq Scan on tenk1 a  (cost=0.00..12.90 rows=290 width=4)
         ->  Hash  (cost=17.62..17.62 rows=200 width=4)
               ->  HashAggregate  (cost=13.62..15.62 rows=200 width=4)
                     ->  Seq Scan on tenk1 b  (cost=0.00..12.90 rows=290 width=4)
(7 rows)
```

Comment (bhuvnesh2703) [inline src/test/regress/expected/subselect_gp.out]: We took a look at the upstream results, there as well we see width as 0.
It looks like if the operator on top of HASHJOIN is not using any projection from HASHJOIN, in that case we are setting the width to 0. 
Plan from postgres 8.4
postgres=# EXPLAIN select count(*) from  (select 1 from tenk1 a where unique1 IN (select distinct hundred from tenk1 b)) ss;
                                    QUERY PLAN
----------------------------------------------------------------------------------
 Aggregate  (cost=37.38..37.39 rows=1 width=0)
   ->  Hash Join  (cost=20.12..37.01 rows=145 width=0)
         Hash Cond: (a.unique1 = b.hundred)
         ->  Seq Scan on tenk1 a  (cost=0.00..12.90 rows=290 width=4)
         ->  Hash  (cost=17.62..17.62 rows=200 width=4)
               ->  HashAggregate  (cost=13.62..15.62 rows=200 width=4)
                     ->  Seq Scan on tenk1 b  (cost=0.00..12.90 rows=290 width=4)
(7 rows)

**Pull request merged**